Generate AppImage

probonopd opened the pull request:
Closes #2024

[AppImage](http://appimage.org/) is a packaging format for Linux that lets users download one single file, set the executable bit and run the application (here: PowerShell). Unlike other solutions, no special runtime infrastructure must be installed on the target system(s).

This pull request, when merged, will package PowerShell as an AppImage and upload it.

probonopd committed:
Update .travis.yml

probonopd committed:
Create appimage.sh

[ci skip]

probonopd committed:
Update travis.sh

probonopd committed:
Use the deb that has been generated in this build

probonopd committed:
Set git clone depth to 100

As per https://github.com/PowerShell/PowerShell/issues/2024#issuecomment-241584449

probonopd committed:
Fix artifacts search path

msftclas commented:
Hi __@probonopd__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
    </span>

TTYL, MSBOT;

msftclas commented:
__@probonopd__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, MSBOT;

probonopd committed:
Do not call appimage.sh from travis.sh but from build.psm1

[ci skip]

probonopd committed:
Call appimage.sh from build.psm1

probonopd committed:
Add AppImage to linux.md

probonopd committed:
Add AppImage to linux.md

probonopd committed:
Add AppImage to linux.md

probonopd committed:
Add AppImage to README.md

probonopd commented:
Here is the log of a build in which the AppImage actually gets generated (it does not in PRs, as do the other packages): https://travis-ci.org/probonopd/PowerShell/builds/154322840

probonopd committed:
"Many" instead of "most", link to AppImage for more info

https://github.com/PowerShell/PowerShell/pull/2027#discussion_r75960080

probonopd commented on docs/installation/linux.md:
```diff
@@ -2,14 +2,15 @@ Package installation instructions
 
 Supports [Ubuntu 14.04][u14], [Ubuntu 16.04][u16],
-[CentOS 7][cos], and [OS X 10.11][osx].
+[CentOS 7][cos], [most Linux distributions (AppImage)][lai], and [OS X 10.11][osx].
```
Yes, done

probonopd committed:
"Many" instead of "most" again

https://github.com/PowerShell/PowerShell/pull/2027#discussion_r75960207

probonopd commented on README.md:
```diff
@@ -27,21 +27,24 @@ You can download and install a PowerShell package for any of the following platf
 | Ubuntu 16.04                   | [.deb][rl-ubuntu16]    | [Instructions][in-ubuntu16]   |
 | Ubuntu 14.04                   | [.deb][rl-ubuntu14]    | [Instructions][in-ubuntu14]   |
 | CentOS 7                       | [.rpm][rl-centos]      | [Instructions][in-centos]     |
+| Most Linux distributions       | [.AppImage][rl-ai]     | [Instructions][in-appimage]   |
```
Done

probonopd committed:
Copyright and license

https://github.com/PowerShell/PowerShell/pull/2027#discussion_r75960505

probonopd committed:
MIT License for appimage.sh

https://github.com/PowerShell/PowerShell/pull/2027#discussion_r75960505

probonopd committed:
Full text of the MIT License is in license_thirdparty_proprietary.txt

probonopd commented on tools/appimage.sh:
```diff
@@ -0,0 +1,143 @@
+
+cd ./$APP/
+
+wget -q https://github.com/probonopd/AppImages/raw/master/functions.sh -O ./functions.sh
```
Done

probonopd committed:
Use icon from local repository

https://github.com/PowerShell/PowerShell/pull/2027#discussion_r75960616

probonopd commented on tools/appimage.sh:
```diff
@@ -0,0 +1,143 @@
+Terminal=true
+EOF
+
+wget -c "https://github.com/PowerShell/PowerShell/raw/master/assets/Powershell_256.png" -O $LOWERAPP.png
```
Done

probonopd commented:
Thanks for the review @TravisEz13.

> Can you do an interactive rebase so that there is one change per file (or less?)

As in
```
git rebase -i ${HASH}
git config --global push.default simple
git push -f
```
Correct?

Wouldn't it be less dangerous if you would use the GitHub [pull request merge squashing](https://help.github.com/articles/about-pull-request-merge-squashing/) setting?

probonopd commented:
Now it says "All checks have passed". Will file an issue regarding the failed upload.

probonopd commented on tools/appimage.sh:
```diff
@@ -0,0 +1,146 @@
+
+# The MIT License (MIT)
+# Copyright (c) 2016 Simon Peter
+
```
Yes, this is based on this [existing script](https://github.com/probonopd/AppImages/blob/master/LICENSE) which comes from the [probonopd/AppImages](https://github.com/probonopd/AppImages/) project which in its entirety [is MIT licensed](https://github.com/probonopd/AppImages/blob/master/LICENSE). Hence I don't think that "copyright Microsoft Corporation" is applicable here, please advise.

probonopd committed:
Clarify license and clean up unused code

https://github.com/PowerShell/PowerShell/pull/2027#discussion_r76104601

andschwa commented:
This looks good to me. Having not used AppImage, I can't comment as to the actual setup, but it clearly works 😄 

@TravisEz13 are you okay with licensing/copyright?

@probonopd do you know what happened at the [end of the log](https://travis-ci.org/probonopd/PowerShell/builds/154322840#L1366) of the build you posted?

andschwa commented:
I'm running a non-PR [build here](https://travis-ci.org/andschwa/PowerShell/builds/155402372) to test.

andschwa commented:
Yeah, can't quite merge. The build failed to run the Pester tests afterwards. @probonopd is it a `cd` issue?

probonopd commented:
If you are referring to [this](https://travis-ci.org/probonopd/PowerShell/builds/154322840#L1366-L1375), what is it supposed to do?

And yes, could be necessary to `cd ..`, will try that.

probonopd committed:
Update appimage.sh

andyleejordan committed:
Move AppImage generation to travis.sh

As it works independently of New-UnixPackage,
I don't think it should be run there.

andyleejordan committed:
Rename AppImage to "powershell" for package consistency

Also put version back to 6.0.0-alpha.9 (- instead of .).
Removed $LOWERAPP as it was no longer needed.

andyleejordan committed:
Clone 1000 commits on Travis CI

The previous failure we saw was because the most recent annotated tag
did not exist in the previous 50 commits. In my experience we've gone
about ~300 commits between tags, so I believe 1000 to be safe, but know
100 will not be.

andschwa commented:
It doesn't appear to be a problem related to the AppImage. I tested both saving and restoring the location; and not doing the AppImage generation until after the Pester tests. I'm testing a rebase now.

andyleejordan committed:
Mark appimage.sh as executable

probonopd committed:
Merge pull request #1 from andschwa/appimage

Suggested changes for AppImage

probonopd commented:
Continued in #2082, hence closing here.
